pandas/core/construction.py

```diff
@@ -510,11 +512,18 @@ def sanitize_masked_array(data: ma.MaskedArray) -> np.ndarray:
     Convert numpy MaskedArray to ensure mask is softened.
     """
     mask = ma.getmaskarray(data)
+    original = data
+    original_dtype = data.dtype
     if mask.any():
         dtype, fill_value = maybe_promote(data.dtype, np.nan)
         dtype = cast(np.dtype, dtype)
         data = ma.asarray(data.astype(dtype, copy=True))
         data.soften_mask()  # set hardmask False if it was True
+        if not mask.all():
+            idx = np.unravel_index(np.nanargmax(data, axis=None), data.shape)
+            if not mask[idx] and int(data[idx]) != original[idx]:
+                if is_integer_dtype(original_dtype) and is_float_dtype(data.dtype):
+                    data = ma.asarray(original, "object")
```

Comment:
I'm not sure if this is the appropriate solution. Converting to `object` should be avoided

Reply:
Could you please expand on your idea? What could I use instead of an object? Thanks for your help!

Reply:
I imagine the `dtype` output from `maybe_promote` isn't entirely appropriate given the original issue. Casting to `object` would cover up the root issue. This issue I suspect is fairly non-trivial.